File: edge/tool-bridge/src/catfish_tool_bridge/email_runtime.py
"""Discover the validated email runtime without importing COM in this daemon."""
import os
from pathlib import Path
import re
import shutil
import sys

from .hermes_paths import hermes_home


def isolated_scripts(root: Path) -> Path | None:
    try:
        generation = (root / 'email-runtime/current.txt').read_text(encoding='utf-8').strip()
    except (OSError, UnicodeError):
        return None
    if not re.fullmatch(r'env-[0-9a-fA-F]{32}', generation):
        return None
    scripts = root / 'email-runtime' / generation / 'Scripts'
    if all((scripts / name).is_file() for name in ('python.exe', 'catfish-email.exe')):
        return scripts
    return None


def find_email_cli() -> str | None:
    explicit = os.environ.get('CATFISH_EMAIL_BIN')
    if explicit and Path(explicit).is_file():
        return explicit
    if sys.platform == 'win32':
        scripts = isolated_scripts(hermes_home())
        if scripts:
            return str(scripts / 'catfish-email.exe')
        legacy = hermes_home() / 'hermes-agent/venv/Scripts/catfish-email.exe'
        if legacy.is_file():
            return str(legacy)
    path = shutil.which('catfish-email')
    if path:
        return path
    local = Path.home() / '.local/bin/catfish-email'
    return str(local) if local.is_file() and os.access(local, os.X_OK) else None

File: edge/companion-app/src-tauri/src/services/email_runtime.rs
//! Resolve the validated isolated Windows email environment.
use std::path::{Path, PathBuf};

pub fn scripts(hermes_home: &Path) -> Option<PathBuf> {
    let root = hermes_home.join("email-runtime");
    let generation = std::fs::read_to_string(root.join("current.txt")).ok()?;
    let generation = generation.trim();
    let suffix = generation.strip_prefix("env-")?;
    if suffix.len() != 32 || !suffix.bytes().all(|c| c.is_ascii_hexdigit()) { return None; }
    let scripts = root.join(generation).join("Scripts");
    if scripts.join("python.exe").is_file() && scripts.join("catfish-email.exe").is_file() {
        Some(scripts)
    } else { None }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn only_published_complete_generation_is_used() {
        let dir = tempfile::tempdir().unwrap();
        let root = dir.path().join("email-runtime");
        let name = "env-0123456789abcdef0123456789abcdef";
        let bin = root.join(name).join("Scripts");
        std::fs::create_dir_all(&bin).unwrap();
        std::fs::write(root.join("current.txt"), name).unwrap();
        assert!(scripts(dir.path()).is_none());
        for exe in ["python.exe", "catfish-email.exe"] { std::fs::write(bin.join(exe), b"test").unwrap(); }
        assert_eq!(scripts(dir.path()), Some(bin));
        std::fs::write(root.join("current.txt"), "../hermes-agent/venv").unwrap();
        assert!(scripts(dir.path()).is_none());
    }
}
